packages/sanity/src/core/form/inputs/files/ImageInput/ImageInput.tsx

```
@@ -239,35 +234,83 @@ function BaseImageInputComponent(props: BaseImageInputProps): React.JSX.Element
+
   const handleStaleUpload = useCallback(() => {
     setIsStale(true)
   }, [])
-  const handleSelectFiles = useCallback(
-    (files: File[]) => {
-      if (directUploads && !readOnly) {
-        uploadFirstAccepted(files)
-      } else if (hoveringFiles.length > 0) {
-        handleFilesOut()
+
+  const handleSelectFilesToUpload = useCallback(
+    (assetSource: AssetSource, files: File[]) => {
+      setSelectedAssetSource(assetSource)
+      const uploader = assetSource.uploader
+      if (uploader) {
+        try {
+          uploaderRef.current = {
+            unsubscribe: uploader.subscribe((event) => {
+              switch (event.type) {
+                case 'progress':
+                  onChange(
+                    PatchEvent.from([
+                      set(Math.max(2, event.progress), [UPLOAD_STATUS_KEY, 'progress']),
+                      set(new Date().toISOString(), [UPLOAD_STATUS_KEY, 'updatedAt']),
+                    ]),
+                  )
+                  break
+                case 'error':
+                  event.files.forEach((file) => {
+                    console.error(file.error)
+                  })
+                  push({
+                    status: 'error',
+                    description: t('asset-sources.common.uploader.upload-failed.description'),
+                    title: t('asset-sources.common.uploader.upload-failed.title'),
+                  })
+                  break
+                case 'all-complete':
+                  onChange(PatchEvent.from([unset([UPLOAD_STATUS_KEY])]))
+                  setSelectedAssetSource(null)
+                  setIsUploading(false)
+                  setMenuOpen(false)
+                  uploaderRef.current = null
+                  break
+                default:
+              }
+            }),
+            uploader: assetSource.uploader,
+          }
+          setIsUploading(true)
+          onChange(PatchEvent.from(createInitialUploadPatches(files[0])))
```

Q: Should we handle if `files[0]` is `undefined`?

---

Should be sufficient just to throw in beginning of the function if files[0] is undefined I think?